File: Adapters/Vungle/ISVungleAdapter/src/main/java/com/ironsource/adapters/vungle/VungleInitializer.java
package com.ironsource.adapters.vungle;

import android.content.Context;

import com.ironsource.mediationsdk.logger.IronLog;
import com.vungle.ads.InitializationListener;
import com.vungle.ads.VungleAds;
import com.vungle.ads.VungleException;

import java.util.ArrayList;
import java.util.concurrent.atomic.AtomicBoolean;

public class VungleInitializer implements InitializationListener {

    private static final VungleInitializer instance = new VungleInitializer();
    private final AtomicBoolean isInitializing = new AtomicBoolean(false);
    private final ArrayList<VungleInitializationListener> initListeners;

    public static VungleInitializer getInstance() {
        return instance;
    }

    private VungleInitializer() {
        initListeners = new ArrayList<>();
        VungleAds.setIntegrationName(
                VungleAds.WrapperFramework.ironsource,
                com.ironsource.adapters.vungle.BuildConfig.VERSION_NAME.replace('.', '_'));
    }

    public void initialize(
            final String appId,
            final Context context,
            VungleInitializationListener listener) {

        if (VungleAds.isInitialized()) {
            listener.onInitializeSuccess();
            return;
        }

        if (isInitializing.getAndSet(true)) {
            initListeners.add(listener);
            return;
        }

        IronLog.ADAPTER_API.verbose("appId = " + appId);

        VungleAds.init(context, appId, this);
        initListeners.add(listener);
    }

    @Override
    public void onSuccess() {
        IronLog.ADAPTER_CALLBACK.verbose("Succeeded to initialize SDK");

        for (VungleInitializationListener listener : initListeners) {
            listener.onInitializeSuccess();
        }

        initListeners.clear();
        isInitializing.set(false);
    }

    @Override
    public void onError(VungleException exception) {
        IronLog.ADAPTER_CALLBACK.verbose(
                "Failed to initialize SDK - " + exception.getLocalizedMessage());

        for (VungleInitializationListener listener : initListeners) {
            listener.onInitializeError(exception.getLocalizedMessage());
        }

        initListeners.clear();
        isInitializing.set(false);
    }

    public interface VungleInitializationListener {

        void onInitializeSuccess();

        void onInitializeError(String error);
    }
}

File: Adapters/Vungle/ISVungleAdapter/src/main/java/com/ironsource/adapters/vungle/VungleInitializer.kt
package com.ironsource.adapters.vungle

import android.content.Context
import com.ironsource.mediationsdk.logger.IronLog
import com.vungle.ads.InitializationListener
import com.vungle.ads.VungleAds
import com.vungle.ads.VungleAds.WrapperFramework
import com.vungle.ads.VungleError
import java.util.concurrent.atomic.AtomicBoolean

class VungleInitializer private constructor() : InitializationListener {
    private val isInitializing = AtomicBoolean(false)
    private val initListeners = mutableListOf<VungleInitializationListener>()

    init {
        VungleAds.setIntegrationName(
            WrapperFramework.ironsource,
            BuildConfig.VERSION_NAME.replace('.', '_')
        )
    }

    fun initialize(
        appId: String,
        context: Context,
        listener: VungleInitializationListener
    ) {
        if (VungleAds.isInitialized()) {
            listener.onInitializeSuccess()
            return
        }
        if (isInitializing.getAndSet(true)) {
            initListeners.add(listener)
            return
        }
        IronLog.ADAPTER_API.verbose("appId = $appId")
        VungleAds.init(context, appId, this)
        initListeners.add(listener)
    }

    override fun onSuccess() {
        IronLog.ADAPTER_CALLBACK.verbose("Succeeded to initialize SDK")
        for (listener in initListeners) {
            listener.onInitializeSuccess()
        }
        initListeners.clear()
        isInitializing.set(false)
    }

    override fun onError(vungleError: VungleError) {
        IronLog.ADAPTER_CALLBACK.verbose(
            "Failed to initialize SDK - " + vungleError.errorMessage
        )
        for (listener in initListeners) {
            listener.onInitializeError(vungleError.errorMessage)
        }
        initListeners.clear()
        isInitializing.set(false)
    }

    interface VungleInitializationListener {
        fun onInitializeSuccess()
        fun onInitializeError(error: String?)
    }

    companion object {
        @JvmField
        val instance = VungleInitializer()
    }
}
